registry/auth/htpasswd/access.go

```diff
@@ -111,6 +119,34 @@ func (ch challenge) Error() string {
 	return fmt.Sprintf("basic authentication challenge for realm %q: %s", ch.realm, ch.err)
 }
 
+// createHtpasswdFile creates and populates htpasswd file with a new user in case the file is missing
+func createHtpasswdFile(path string) error {
+	if _, err := os.Open(path); os.IsNotExist(err) {
+		if err := os.MkdirAll(filepath.Dir(path), 0700); err != nil {
+			return err
+		}
+		f, err := os.OpenFile(path, os.O_WRONLY|os.O_CREATE, 0600)
+		if err != nil {
+			return fmt.Errorf("failed to open htpasswd path %s", err)
+		}
+		defer f.Close()
+		var secretBytes [32]byte
+		if _, err := rand.Read(secretBytes[:]); err != nil {
+			return err
+		}
+		pass := base64.URLEncoding.EncodeToString(secretBytes[:])
+		encryptedPass, err := bcrypt.GenerateFromPassword([]byte(pass), bcrypt.DefaultCost)
+		if err != nil {
+			return err
+		}
+		if _, err := f.Write([]byte(fmt.Sprintf("docker:%s", string(encryptedPass[:])))); err != nil {
+			return err
+		}
+		logrus.Warnf("htpasswd is missing. provisioned with default user:docker password: %s", pass)
```

Comment:
This output looks kind of off
```
WARN[0000] htpasswd is missing. provisioned with default user:docker password: qakOn184C_WYgnUt7RDyE3xO7YrDkYhpqA9A2JbwIz0=
```

Maybe just getting rid of the space after `password: ` would make it look more consistent. Also please use `RawURLEncoding` to get rid of the `=` at the end, some might find it confusing as to whether it is part of the password or part of an encoding.